site/_shortcodes/Example.js

```diff
@@ -17,11 +17,186 @@
 const fs = require('fs/promises');
 const path = require('path');
 const Prism = require('prismjs');
-const md = require('markdown-it')();
+const espree = require('espree');
+const execa = require('execa');
 
 const examples = [];
 let counter = {};
 
+class Example {
+  constructor(nunjucksEngine, context, raw) {
+    this.renderer = nunjucksEngine;
+
+    this.context = context;
+    this.raw = raw;
+
+    this.title = context.title;
+    this.slug = context.page.fileSlug;
+    counter[this.slug] = counter[this.slug] + 1 || 0;
+    this.id = `${this.slug}-${counter[this.slug]}`;
+
+    this.parsed = {
+      webComponents: this._parseHtml(),
+      react: this._parseJavaScript(),
+    }
+
+    this.previews = this._renderPreviews();
+    this.widget = this._renderWidget();
+  }
+
+  _extract(regExp, string) {
+    const match = regExp.exec(string);
+    return match && match[1] ? match[1] : null;
+  }
+
+  _parseCode(language, string) {
+    const code = {
+      all: this._extract(new RegExp('```' + language + '\n(.*?)\n```', 'gms'), string),
+    };
+
+    return code;
+  }
+
+  _parseHtml() {
+    const html = this._parseCode('html', this.raw);
+    if (!html.all) {
+      return null;
+    }
+
+    html.head = this._extract(/<head>(.*?)<\/head>/gms, html.all);
+    html.body = this._extract(/<body>(.*?)<\/body>/gms, html.all);
+
+    // Make sure head and body both don't contain any blank lines
+    // as they would create empty paragraphs in markdown
+    html.head = html.head.replace(/^\s*\n/gm, '').replace(/\s*$/, '');
+    html.body = html.body.replace(/^\s*\n/gm, '').replace(/\s*$/, '');
+
+    return html;
+  }
+
+  _parseJavaScript() {
+    const js = this._parseCode('javascript', this.raw);
+    if (!js.all) {
+      return null;
+    }
+
+    let ast = {};
+    try {
+      ast = espree.parse(js.all, {
+        ecmaVersion: 'latest',
+        sourceType: 'module',
+        ecmaFeatures: {
+          jsx: true,
+        }
+      });
+    } catch (e) {
+      console.error(`[Examples] Could not retrieve AST for ${this.fileSlug}`);
+      return null;
+    }
+
+    js.imports = ast.body.filter((node) => {
+      return node.type === 'ImportDeclaration'
+    }).map((node) => {
+      return js.all.substring(node.start, node.end);
+    });
+
+    js.app = ast.body.find((node) => {
+      return node.type === 'FunctionDeclaration'
+    });
+
+    if (!js.app) {
+      console.warn(`[Examples] No function in ${this.fileSlug}`);
+      return null;
+    }
+
+    js.appName = js.app.id.name;
+    js.app = js.all.substring(js.app.start, js.app.end);
+
+    return js;
+  }
+
+  _renderPreviews() {
+    const previews = {};
+    if (this.parsed.webComponents) {
```

Comment:
It feels like it'd be better to extract the JS and wc parsing/rendering logic into separate classes to avoid these repeated if statements. Or is there a case when both are valid at the same time?

Reply:
Yes that would be cleaner. When building I always had in mind that at some point one might want to extend this to be able to power a widget where you are able to toggle between web component and react sample like we had in mind for the hero example. That's when you would have both a `js` and a `html` snippet in one `{% example %}`. But that's probably not a valid use case anymore...?

Reply:
The hero examples are standalone anyway.